### PR TITLE
Updating our BatchSamplers to use the same scale across all replicas.

### PR DESCRIPTION
**Pull request by alxmrs**
This updates both our equivalence batch samplers to make it that all n_workers / num_replicas have batches with the same scale of data. This is needed because in DDP training across n_workers, when we mix scales, we're liable to NCCL timeouts. These timeouts are stochastic; they're caused by the scenario where we load 1/4 scale on at least one worker, but 1 or 1/2 scale on many of the other workers. When batch norm is turned on, and thus we need to synchronize between all workers, we have a high likelihood of facing GPU level timeouts, since data loading can take so much longer for larger datasets. 

This also fixes a small, but significant, bug where we didn't actually set epochs properly!

### CODE REVIEW

**Comment by alxmrs** [comment]
@codex may I have a review?

**Comment by alxmrs** [comment]
@codex may I have your review?

**Comment by alxmrs** [comment]
Thanks @codex. May I have another review?

File: src/ocean_emulators/train.py [hunk 1]
```diff
@@ -415,9 +415,9 @@ def run(self) -> None:
                 cur_step = self.get_current_step(epoch)
                 self.init_data_loaders(cur_step)
 
-            if isinstance(self.train_sampler, DistributedSampler):
+            if hasattr(self.train_sampler, "set_epoch"):
                 self.train_sampler.set_epoch(epoch)
-            if isinstance(self.val_sampler, DistributedSampler):
+            if hasattr(self.val_sampler, "set_epoch"):
                 self.val_sampler.set_epoch(epoch)
```
Comment (alxmrs) [inline src/ocean_emulators/train.py:440]: Bugfix: we weren't setting the epochs in the distributed sampler!

**Comment by alxmrs** [comment]
Thanks, @codex. May I have another review?

**Comment by alxmrs** [comment]
Good shout, @codex. may I have another review?

**Comment by alxmrs** [comment]
Thanks again, @codex. May I have one last review?

File: src/ocean_emulators/utils/samplers.py [hunk 2]
```diff
@@ -234,29 +263,52 @@ def set_epoch(self, epoch: int) -> None:
...
+        # data in the same training step, which causes load-time asymmetry and
+        # Gloo/NCCL timeouts.
+        chunks: list[tuple] = []
+        for sampler in self._inner._samplers:
+            group_chunks = list(itertools.batched(sampler, self.num_replicas))
+            if group_chunks and len(group_chunks[-1]) < self.num_replicas:
+                # Always pad incomplete chunks so every chunk is homogeneous
+                # and full-sized. Without padding, an incomplete chunk would
+                # break the stride alignment for all subsequent chunks.
+                # Dropping here would erase entire groups that have fewer
```
[the remark this answers is not in the document]
Comment (alxmrs) [reply src/ocean_emulators/utils/samplers.py]: This is a contradiction with the previous comment. Keeping things as they are for now.

File: src/ocean_emulators/utils/samplers.py [hunk 3]
```diff
@@ -142,27 +150,47 @@ def from_datasets(
...
-        batch_sampler = itertools.chain(*self._samplers)
+        # Batch within each group and align to num_replicas boundaries so that
+        # consecutive num_replicas batches always belong to the same group.
+        # This prevents different DDP ranks from receiving different-resolution
+        # data in the same training step, which causes load-time asymmetry and
+        # Gloo/NCCL timeouts.
+        per_replica = []
+        move_to_end = []
+        for sampler in self._samplers:
+            pre_replica_per_sample = list(itertools.batched(sampler, self.num_replicas))
```
Comment (jder) [inline src/ocean_emulators/utils/samplers.py:164]: Is this supposed to be `per_replica_per_sample` as opposed to `pre...`?

**Review by jder — COMMENTED** [review]
I was finding it hard to understand the logic here and ended up feeling like it was strange to have num_replicas at all in the non-distributed sampler. I ended up with an alternative and I think simpler option in #717 -- WDYT?

**Comment by jder** [comment]
Closing in favor of #717
